libraries/aws/shadow/src/aws_iot_shadow_api.c

```diff
@@ -338,121 +353,123 @@ static AwsIotShadowError_t _setCallbackCommon( IotMqttConnection_t mqttConnectio
-                                                   IotMqtt_SubscribeSync );
-
-            if( status == AWS_IOT_SHADOW_SUCCESS )
+            /* Replace existing callback. */
+            if( pCallbackInfo != NULL )
             {
+                IotLogInfo( "(%.*s) Found existing %s callback. Replacing callback.",
+                            thingNameLength,
+                            pThingName,
+                            _pAwsIotShadowCallbackNames[ type ] );
+
                 pSubscription->callbacks[ type ] = *pCallbackInfo;
             }
+            /* Remove existing callback. */
             else
             {
-                /* On failure, check if this subscription can be removed. */
+                IotLogInfo( "(%.*s) Removing existing %s callback.",
+                            thingNameLength,
+                            pThingName,
+                            _pAwsIotShadowCallbackNames[ type ] );
+
+                /* Unsubscribe, then clear the callback information. */
+                ( void ) _modifyCallbackSubscriptions( mqttConnection,
+                                                       type,
+                                                       pSubscription,
+                                                       IotMqtt_UnsubscribeSync );
+                ( void ) memset( &( pSubscription->callbacks[ type ] ),
+                                 0x00,
+                                 sizeof( AwsIotShadowCallbackInfo_t ) );
+
+                /* Check if this subscription object can be removed. */
                 _AwsIotShadow_RemoveSubscription( pSubscription, NULL );
             }
         }
-        /* Do nothing; set return value to success. */
+        /* No existing callback. */
         else
         {
-            status = AWS_IOT_SHADOW_SUCCESS;
+            /* Add new callback. */
+            if( pCallbackInfo != NULL )
+            {
+                IotLogInfo( "(%.*s) Adding new %s callback.",
+                            thingNameLength,
+                            pThingName,
+                            _pAwsIotShadowCallbackNames[ type ] );
+
+                status = _modifyCallbackSubscriptions( mqttConnection,
+                                                       type,
+                                                       pSubscription,
+                                                       IotMqtt_SubscribeSync );
+
+                if( status == AWS_IOT_SHADOW_SUCCESS )
+                {
+                    pSubscription->callbacks[ type ] = *pCallbackInfo;
+                }
+                else
+                {
+                    /* On failure, check if this subscription can be removed. */
+                    _AwsIotShadow_RemoveSubscription( pSubscription, NULL );
+                }
+            }
+            /* Do nothing; set return value to success. */
+            else
+            {
+                status = AWS_IOT_SHADOW_SUCCESS;
```

Comment:
This block seems unnecessary since status must be AWS_IOT_SHADOW_SUCCESS at this point

Reply:
Agree. Removed this else block.